### PR TITLE
added remove to iterable map

### PR DESCRIPTION
**Pull request by joeykrug**
﻿# Pull Request Template
### Examples (if necessary)

### CODE REVIEW

File: src/libraries/iterableMap.se [hunk 1]
```diff
@@ -49,3 +57,6 @@ def contains(key):
 
 def count():
     return self.numberOfItems
+
+def getOffsetByKey(key):
+    return self.itemsMap[key].offset
```
Comment (MicahZoltu) [inline src/libraries/iterableMap.se]: Recommend not exposing this.  It is an implementation detail and shouldn't be relied on by users.

File: src/libraries/iterableMap.se [hunk 2]
```diff
@@ -28,7 +29,14 @@ def addOrUpdate(key, value):
     else:
         self.update(key, value)
 
-# TODO: add support for item removal, note that this requires moving the last item in the collection to the newly opened slot so we don't end up with a sparse collection
+def remove(key):
+    require(self.contains(key))
+    # move last item in collection to the newly opened slot
+    self.itemsArray[self.itemsMap[key].offset] = self.getByOffset(self.numberOfItems)
```
Comment (MicahZoltu) [inline src/libraries/iterableMap.se]: What happens if item being removed is the last item?  What happens if item being removed is the only item?  I haven't thought through these, so maybe this code works with them already, but wanted to make sure someone thinks about it.

File: src/libraries/iterableMap.se (hunk 1 above)
Comment (joeykrug) [inline src/libraries/iterableMap.se]: Fixing in next commit

**Comment by joeykrug** [comment]
Yep, there were some I found when testing, I think they're fixed in latest commit

File: src/libraries/iterableMap.se [hunk 3]
```diff
@@ -28,7 +29,20 @@ def addOrUpdate(key, value):
...
 
-# TODO: add support for item removal, note that this requires moving the last item in the collection to the newly opened slot so we don't end up with a sparse collection
+def remove(key):
+    require(self.contains(key))
+    keyRemovedOffset = self.itemsMap[key].offset
+    self.itemsArray[keyRemovedOffset] = 0
+    self.itemsMap[key].hasValue = 0
+    self.itemsMap[key].value = 0
+    self.itemsMap[key].offset = 0
+    if(self.numberOfItems > 1 and keyRemovedOffset != self.numberOfItems):
```
Comment (MicahZoltu) [inline src/libraries/iterableMap.se]: Won't the second half of this always be true?  Number of items will be one greater than the offset of the last item, meaning in will never equal offset of removed item.

Comment (joeykrug) [inline src/libraries/iterableMap.se]: Yeah good point, fixing

**Pull request merged**